LoopKitUI/View Controllers/ServiceTableViewController.swift

```diff
@@ -0,0 +1,154 @@
+//
+//  ServiceTableViewController.swift
+//  LoopKitUI
+//
+//  Created by Darin Krauss on 5/23/19.
+//  Copyright © 2019 LoopKit Authors. All rights reserved.
+//
+
+import LoopKit
+
+open class ServiceTableViewController: UITableViewController {
+
+    public enum Operation {
+        case create
+        case update
+    }
+
+    public let service: Service
+    public let operation: Operation
+
+    public init(service: Service, for operation: Operation) {
+        self.service = service
+        self.operation = operation
+
+        super.init(style: .grouped)
+    }
+
+    public required init?(coder: NSCoder) {
+        fatalError("init(coder:) has not been implemented")
+    }
+
+    open override func viewDidLoad() {
+        super.viewDidLoad()
+
+        title = service.localizedTitle
+
+        if operation == .create {
+            navigationItem.leftBarButtonItem = UIBarButtonItem(barButtonSystemItem: .cancel, target: self, action: #selector(cancel))
+        }
+        navigationItem.rightBarButtonItem = UIBarButtonItem(barButtonSystemItem: .done, target: self, action: #selector(done))
+
+        updateButtonStates()
+    }
+
+    public func updateButtonStates() {
+        navigationItem.rightBarButtonItem?.isEnabled = service.hasValidConfiguration
+    }
+
+    @objc public func cancel() {
+        view.endEditing(true)
+
+        notifyComplete()
+    }
+
+    @objc public func done() {
+        guard service.hasValidConfiguration else {
+            return
+        }
+
+        view.endEditing(true)
+
+        UIView.animate(withDuration: 0.25, animations: {
+            self.navigationItem.titleView = ValidatingIndicatorView(frame: CGRect.zero)
+        })
+
+        service.verifyConfiguration { error in
+            DispatchQueue.main.async {
+                UIView.animate(withDuration: 0.25, animations: {
+                    self.navigationItem.titleView = nil
+                })
+
+                if let error = error {
+                    self.present(UIAlertController(with: error), animated: true)
+                    return
+                }
+
+                switch self.operation {
+                case .create:
+                    self.notifyCreated()
+                case .update:
+                    self.notifyUpdated()
+                }
+            }
+        }
+    }
+
+    public func confirmDeletion(completion: (() -> Void)? = nil) {
+        view.endEditing(true)
+
+        let alert = UIAlertController(serviceDeletionHandler: {
+            self.notifyDeleted()
+        })
+
+        present(alert, animated: true, completion: completion)
+    }
+
+    private func notifyCreated() {
+        service.notifyCreated {
```

Comment:
I now see this is probably the reason you've added these methods to the Service protocol.  They make more sense as part of a ServiceTableViewControllerDelegate protocol.  Not all services will use this tableview, I imagine.  It's fine to have this VC here as a convenience, but it should have its own delegate, and the methods for it not mixed in with the Service protocol.  That allows it to be used independently.

Reply:
I believe I have straightened out the notification and delegate issues now. Take a look and let me know what you think.